### PR TITLE
fix: better input to bignumber handling

### PR DESCRIPTION
**Pull request by chrisling-dev**
https://www.notion.so/prepo/FE-Project-Management-d3878c29e9534234a50f4f8838525bee?p=80c28d3226fc47cdbe8dace9694b6522&pm=s

### CODE REVIEW

File: packages/prepo-utils/src/validateStringToBN.ts [hunk 1]
```diff
@@ -1,15 +1,4 @@
-import { parseUnits } from 'ethers/lib/utils'
-import { getDecimals } from './getDecimals'
-import { safeStringBN } from './safeStringBN'
-
-export const validateStringToBN = (input: string, decimals?: number): boolean => {
-  if (decimals === undefined) return false
+export const validateStringToBN = (input: string): boolean => {
   if (input === '') return true
```
Comment (XavierEkkel) [inline packages/prepo-utils/src/validateStringToBN.ts:2]: add comment that this is because we treat '' as 0

File: packages/prepo-utils/jest.config.ts [hunk 2]
```diff
@@ -7,7 +7,6 @@ const fromRoot = (d): string => path.join(__dirname, d)
 const config: Config.InitialOptions = {
   ...base,
   roots: [fromRoot('.')],
-  name: 'prepo-utils',
```
Comment (XavierEkkel) [inline packages/prepo-utils/jest.config.ts:10]: why?

File: apps/frontend/core-dapp/src/stores/entities/Erc20.entity.ts [hunk 3]
```diff
@@ -229,11 +229,9 @@ export class Erc20Store extends ContractStore<RootStore, SupportedContracts> {
     spenderContractName: SupportedContractsNames = 'UNISWAP_SWAP_ROUTER'
   ): boolean | undefined {
     const contractAddresses = supportedContracts[spenderContractName]
-    if (!contractAddresses || this.decimalsNumber === undefined) return undefined
-    return this.needsToAllowTokens(
-      contractAddresses[this.root.web3Store.network.name],
-      parseUnits(safeStringBN(amount), this.decimalsNumber)
-    )
+    const amountBN = parseUnits(amount, this.decimalsNumber)
+    if (!contractAddresses || amountBN === undefined) return undefined
```
Comment (XavierEkkel) [inline apps/frontend/core-dapp/src/stores/entities/Erc20.entity.ts:233]: add undefined check for `contractAddresses[this.root.web3Store.network.name]` as well?

File: packages/prepo-utils/src/parseUnits.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+): BigNumber | undefined => {
+  // loading decimals of ERC20 token
+  if (decimals === undefined) return undefined
+
+  // force value to be string for rest of the logic to work
+  let valueString = `${value}`
+
+  // prevent ether's parseUnits from throwing error during runtime
+  // when parsing non number input
+  if (Number.isNaN(+valueString)) return undefined
```
Comment (XavierEkkel) [inline packages/prepo-utils/src/parseUnits.ts:19]: use `validateStringToBN` instead?

File: packages/prepo-utils/src/parseUnits.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+  if (valueString === '') return BigNumber.from(0)
+
+  const [significantPart, fractionPart] = valueString.split('.')
+
+  // cap fraction part within `decimals` because
```
Comment (XavierEkkel) [inline packages/prepo-utils/src/parseUnits.ts:25]: nit:
```suggestion
  // cap fraction part within `decimals` (not rounded up) because
```

File: packages/prepo-utils/__tests__/utils.test.ts [hunk 6]
```diff
@@ -37,3 +38,45 @@ describe('prepo-utils', () => {
...
   })
 })
+
+describe('parseUnits function', () => {
+  it('returns undefined if decimals is undefined', () => {
+    const output = parseUnits('100', undefined)
+    expect(output).toBe(undefined)
+  })
+
+  it('returns zero BigNumber if string is empty', () => {
```
Comment (XavierEkkel) [inline packages/prepo-utils/__tests__/utils.test.ts:48]: nit:
```suggestion
  it('returns 0 BigNumber if empty string', () => {
```

File: packages/prepo-utils/__tests__/utils.test.ts [hunk 7]
```diff
@@ -37,3 +38,45 @@ describe('prepo-utils', () => {
...
+    const output = parseUnits('0.1234567890', 18)
+    expect(output.eq(BigNumber.from('123456789000000000'))).toBe(true)
+  })
+
+  it('returns undefined if fraction part contains non-number character', () => {
+    const output = parseUnits('0.123456789123456789123456789abc', 18)
+    expect(output).toBe(undefined)
+  })
+
+  it('returns undefined without throwing error if value is invalid number', () => {
```
Comment (XavierEkkel) [inline packages/prepo-utils/__tests__/utils.test.ts:78]: (not throwing error is implied)
```suggestion
  it('returns undefined if value is invalid number', () => {
```

File: packages/prepo-utils/__tests__/utils.test.ts [hunk 8]
```diff
@@ -37,3 +38,45 @@ describe('prepo-utils', () => {
     expect(formattedNumber3).toEqual('$123M')
   })
 })
+
+describe('parseUnits function', () => {
+  it('returns undefined if decimals is undefined', () => {
```
Comment (XavierEkkel) [inline packages/prepo-utils/__tests__/utils.test.ts:43]: nit:
```suggestion
  it('returns undefined if decimals undefined', () => {
```

File: packages/prepo-utils/__tests__/utils.test.ts [hunk 9]
```diff
@@ -37,3 +38,45 @@ describe('prepo-utils', () => {
     expect(formattedNumber3).toEqual('$123M')
   })
 })
+
+describe('parseUnits function', () => {
```
Comment (XavierEkkel) [inline packages/prepo-utils/__tests__/utils.test.ts:42]: nit:
```suggestion
describe('# parseUnits', () => {
```

Comment (XavierEkkel) [inline packages/prepo-utils/__tests__/utils.test.ts:42]: should cover more edge case conditions

File: packages/prepo-utils/__tests__/utils.test.ts [hunk 10]
```diff
@@ -37,3 +38,45 @@ describe('prepo-utils', () => {
...
+    const output = parseUnits('100', undefined)
+    expect(output).toBe(undefined)
+  })
+
+  it('returns zero BigNumber if string is empty', () => {
+    const output = parseUnits('', 6)
+    expect(output.eq(BigNumber.from(0))).toBe(true)
+  })
+
+  it('returns 100000000 if value is 100 in number format and decimals is 6', () => {
```
Comment (XavierEkkel) [inline packages/prepo-utils/__tests__/utils.test.ts:53]: generally use `=` rather than `is`

File: packages/prepo-utils/jest.config.ts (hunk 2 above)
Comment (chrisling-dev) [reply packages/prepo-utils/jest.config.ts:10]: `name` isn't meant to be part of jest config I believe, even their documentation doesn't show a "name" option. And when running yarn test, cli complaints that this is unknown option

Screenshot 2023-01-26 at 10 22 46 AM

File: apps/frontend/core-dapp/src/stores/entities/Erc20.entity.ts (hunk 3 above)
Comment (chrisling-dev) [reply apps/frontend/core-dapp/src/stores/entities/Erc20.entity.ts:233]: don't have to because the function that takes `contractAddresses[this.root.web3Store.network.name]` accepts it as undefined and already handles undefined case there

File: packages/prepo-utils/src/parseUnits.ts [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+/**
+ * @returns BigNumber if input is valid and decimals is loaded
+ * @description Safely converts any string or number to BigNumber without risk of throwing run time error
+ */
+export const parseUnits = (value: string, decimals: number | undefined): BigNumber | undefined => {
+  // loading decimals of ERC20 token
+  if (decimals === undefined) return undefined
+
+  // force value to be string for rest of the logic to work
+  let valueString = value
```
Comment (XavierEkkel) [inline packages/prepo-utils/src/parseUnits.ts:13]: since `value` is passed in only as a string now, is this needed anymore?

File: packages/prepo-utils/__tests__/utils.test.ts [hunk 12]
```diff
@@ -37,3 +45,72 @@ describe('prepo-utils', () => {
...
+
+  // test integer + fraction part together
+  it('returns correct BigNumber if valid full input', () => {
+    validFullInput.forEach(({ input, decimals, output }) => {
+      const result = parseUnits(input, decimals)
+      expect(result.eq(output)).toBe(true)
+    })
+  })
+
+  it('returns undefined if value 12,34', () => {
```
Comment (XavierEkkel) [inline packages/prepo-utils/__tests__/utils.test.ts:107]:
```suggestion
  it('returns undefined if comma in integer part', () => {
```
and group with integer part tests

File: packages/prepo-utils/__tests__/utils.test.ts [hunk 13]
```diff
@@ -37,3 +45,72 @@ describe('prepo-utils', () => {
...
+    const output = parseUnits('12,34', 6)
+    expect(output).toBe(undefined)
+  })
+
+  it('returns undefined if value 12a.34b', () => {
```
Comment (XavierEkkel) [inline packages/prepo-utils/__tests__/utils.test.ts:112]:
```suggestion
  it('returns undefined if non-numbers in integer and fraction parts', () => {
```

File: packages/prepo-utils/__tests__/utils.test.ts [hunk 14]
```diff
@@ -37,3 +45,72 @@ describe('prepo-utils', () => {
...
+  })
+
+  it('returns correct BigNumber if valid integer only input', () => {
+    validIntegerInput.forEach(({ input, decimals, output }) => {
+      const result = parseUnits(input, decimals)
+      expect(result.eq(output)).toBe(true)
+    })
+  })
+
+  it('returns undefined if invalid integer number', () => {
```
Comment (XavierEkkel) [inline packages/prepo-utils/__tests__/utils.test.ts:76]:
```suggestion
  it('returns undefined if non-number in integer part', () => {
```

File: packages/prepo-utils/__tests__/utils.test.ts [hunk 15]
```diff
@@ -37,3 +45,72 @@ describe('prepo-utils', () => {
     expect(formattedNumber3).toEqual('$123M')
   })
 })
+
+describe('# parseUnits', () => {
+  // test decimals
```
Comment (XavierEkkel) [inline packages/prepo-utils/__tests__/utils.test.ts:50]: instead of grouping via comments, can group via a describe block so it will also reflect in the test output via indents

File: packages/prepo-utils/__tests__/utils.test.ts [hunk 16]
```diff
@@ -37,3 +45,72 @@ describe('prepo-utils', () => {
...
+  })
+
+  it('returns undefined if invalid integer number', () => {
+    const output = parseUnits('abcd1234', 6)
+    expect(output).toBe(undefined)
+  })
+
+  // test fraction only value
+  it('returns correct BigNumber if valid fraction only input', () => {
+    validFractionInput.forEach(({ input, decimals, output }) => {
```
Comment (XavierEkkel) [inline packages/prepo-utils/__tests__/utils.test.ts:83]: test cases shouldn't be looped, they should be in single purpose isolated test cases, to avoid a single failure causing all to fall (and making it hard to determine what went wrong)

File: packages/prepo-utils/__mocks__/parseUnitsInput.mock.ts [hunk 17]
```diff
@@ -0,0 +1,93 @@
+import { BigNumber } from 'ethers'
+import { parseEther } from 'ethers/lib/utils'
+
```
Comment (XavierEkkel) [inline packages/prepo-utils/__mocks__/parseUnitsInput.mock.ts:3]: don't need this file, based on other comment about not looping cases within a single test

**Review by XavierEkkel — APPROVED** [review]
LGTM after comments

File: packages/prepo-utils/src/parseUnits.ts (hunk 11 above)
Comment (chrisling-dev) [reply packages/prepo-utils/src/parseUnits.ts:13]: yeap, value cannot be reassigned because it's a parameter, but we need to reassign the valueString if fraction digits is longer than decimals, so yea this is still needed

Comment (chrisling-dev) [reply packages/prepo-utils/src/parseUnits.ts:13]: ahh except the comment is misleading now, will remove
